Automattic/pocket-casts-android
# Onboarding Flow: Initial upgrade screen

| 📘 Project: #524  |
| --- |

### Description
Adds the initial upgrade screen to the onboarding flow.

I ran into a number of issues on this page. I think I've worked through each of them reasonably well in this PR, but I hope you all might have some ideas/suggestions for better ways to address them.

#### 1. Blurred yellow background

Jetpack compose has some really nice blurring functionality, and it works great... on Android 12 and above. For that reason, I pulled the png assets from Figma, but they were really large. For example, the xxxhdpi png alone was over 2 megabytes. I worked on reducing the size and was able to cut it down to about 1/10th of its original size using tinypng.com (everything I tried locally gave me a lower quality image at a larger size). This seems pretty good, but I see just a bit of banding due to the compression.

Therefore, I kept the Jetpack Compose version for Android 12 and 13. Eventually, that will let us drop the png assets and reduce the apk size.

#### 2. Animating the scrolling row with infinite scroll

The way I'm animating the scrolling row seems to work pretty well, but there's just a little bit of stuttering in the animation, even on release builds. I know compose still has some performance issues around lazy lists, so that's probably just a consequence of this, but let me know if you can see any ways to improve this.

I am doing some pretty hacky stuff in this `LazyRow` (see point 3 below), but even without that I got the same stuttering.

In order to get an "infinite" list, I'm using the `LazyRow` to load the relevant items over and over again, but only up to around 200 times, at that point, I just jump back to the beginning of the list. I also tested this going up to `Int.MAX_VALUE`, and that worked fine as well, but I thought it might be a bit safer to not let that list get too crazy large since it's unlikely anyone will ever get to even 100 items.

I tried doing this approach with a regular `Row` instead of a `LazyRow` (in order to avoid the hackiness in point 3 below), but the performance there got noticeably worse around 30 or so items.

#### 3. Variable height row items

In the scrolling row, the items need to be able to accommodate different heights (this becomes an issue if you bump up the font size). In essence, I need to know what the largest item is and then force the entire row (and all the items in it) to be that height. Fortunately, compose has the wonder `IntrisinsicSize.Max` height modifier. Unfortunately, that Modifier can't be used with a `LazyRow`. It can be used with a regular `Row` though, so I'm going with the incredibly hacky approach of including a single `Row` as the first item in the `LazyRow`. I apply the `IntrinsicSize.Max` modifier to the `Row`, then get the `Row`'s height, and apply that height to the `LazyRow`. 

This seems to work pretty well, but I'm not proud of it and I'll be delighted if you have any better ideas.

#### 4. Talkback

When animating the row of features, for some reason Talkback becomes unusable due to lag (I was seeing 30 seconds to 1 minute of delay between me making a gesture and Talkback responding to it on my test device).

I wasn't able to figure out what is the cause of the issue, so I worked around it by checking to see if Talkback is enabled and, if it is enabled, turning off the animation and instead allowing the user to manually scroll the feature row. I actually think this is probably a better experience for a user using touch exploration, but let me know what you think.

### Testing Instructions

1. Make sure that you are not logged into an account, that you have turned on the onboarding feature in the `base.gradle` file, and I suggest you update the [onboarding nav controller's start destination](https://github.com/Automattic/pocket-casts-android/blob/3ccb40f391e01284c8144268830a848d6390eb12/modules/features/account/src/main/java/au/com/shiftyjelly/pocketcasts/account/onboarding/OnboardingFlowComposable.kt#L44) to be `OnboardingNavRoute.plusFeatures` so that you don't have to go through the onboarding flow to get to the upgrade screen each time.
2. If you haven't hardcoded the onboarding flow to go straight to the plus upgrade page, use the onboarding flow to login to an account and skip through the placeholder recommendations page.
3. ✅ Observe that the event `onboarding_upgrade_shown` is fired
4. ✅ Observe that the UI matches the designs
5. ✅ Observe that the row of features is scrolling "infinitely" on its own, that all the items are the same height, and that you cannot scroll the list manually.
6. Turn on TalkBack
7. ✅  Observe that Talkback works as expected, that the list stops scrolling automatically, and that the list now only has one of each item.
8. Turn TalkBack off
9. ✅ Observe that the feature row starts automatically scrolling again and that it appears to be an infinite row

## Screenshots or Screencast 

![image](https://user-images.githubusercontent.com/4656348/202000206-62d8120a-c4ca-4871-a0a5-1edd507bd9a7.png)

## Checklist
 
#### I have tested any UI changes...
